### PR TITLE
Token based Authentication

### LINKED ISSUE #3: Authentication
+ Sign in and sign up APIs 
+ Authorization of user info before accessing every other API
+ Sending `401` for unauthorized users

### PR DESCRIPTION
**Pull request by AnirudhGoel**
fixes #3

### CODE REVIEW

File: api/views.py [hunk 1]
```diff
@@ -1 +1,40 @@
...
+
+    try:
+        user=User.objects.create_user(username, email, password)
+        user.first_name = fName
+        user.last_name = lName
+        user.is_superuser=False
+        user.is_staff=False
+        user.save()
+    except Exception as e:
+        return Response(status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```
Comment (prabhakar267) [inline api/views.py]: Never return 500 response code intentionally from server

File: api/views.py [hunk 2]
```diff
@@ -1 +1,40 @@
...
+
+    if request.method == 'POST':
```
Comment (prabhakar267) [inline api/views.py]: Check seems useless

File: api/views.py [hunk 3]
```diff
@@ -1 +1,40 @@
...
+    """
+
+    fName = request.POST['fName']
+    lName = request.POST['lName']
+    email = request.POST['email']
+    username = request.POST['username']
+    password = request.POST['password']
+
+    try:
+        user=User.objects.create_user(username, email, password)
```
Comment (prabhakar267) [inline api/views.py]: You can use something like `u = User(email=..., pass=...)`

Reference: https://docs.djangoproject.com/en/dev/topics/db/queries/#creating-objects

File: api/views.py [hunk 4]
```diff
@@ -1 +1,40 @@
...
+
 from api.modules.city import view as city
+
+@api_view(['POST'])
+@permission_classes((AllowAny, ))
+def signUp(request):
+    """
+    Adds a new user to Database
+    :param request:
+    :param fname:
```
Comment (prabhakar267) [inline api/views.py]: "Params" are all the variables which are present as method parameters
For object structure of any one param, you can use it's description to elaborate more

File: api/views.py [hunk 5]
```diff
@@ -1 +1,40 @@
...
+from rest_framework.permissions import AllowAny
+from rest_framework.response import Response
+from rest_framework import permissions, status
+from django.contrib.auth.models import User
+
 from api.modules.city import view as city
+
+@api_view(['POST'])
+@permission_classes((AllowAny, ))
+def signUp(request):
```
Comment (prabhakar267) [inline api/views.py]: Use PEP8 style method naming
Reference: https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables

File: api/views.py [hunk 6]
```diff
@@ -1 +1,40 @@
...
+    :param request:
+    :param fname:
+    :param lname:
+    :param email:
+    :param username:
+    :param password:
+    :return:
+    """
+
+    fName = request.POST['fName']
```
Comment (prabhakar267) [inline api/views.py]: Use `firstname` and `lastname`

File: api/views.py (hunk 5 above)
Comment (prabhakar267) [inline api/views.py]: Also add a check for existing user and send a message accordingly, something like `User with email "xx" already present`

**Review by prabhakar267 — CHANGES_REQUESTED** [review]

File: api/views.py (hunk 3 above)
Comment (AnirudhGoel) [reply api/views.py]: Kindly refer to this - https://stackoverflow.com/a/18504852/6229821
User.object.create_user() takes care of setting the password encryption properly.

**Review by prabhakar267 — APPROVED** [review]

**Comment by prabhakar267** [comment]
Squash these changes into one commit and update the PR

**Pull request merged**